Source/Core/DolphinQt/SkylanderPortal/SkylanderPortalWindow.cpp

```diff
@@ -0,0 +1,860 @@
+                           QMessageBox::Ok);
+      return;
+    }
+    m_last_skylander_path = QFileInfo(m_file_path).absolutePath() + QString::fromStdString("/");
+    accept();
+  });
+
+  connect(btn_cancel, &QAbstractButton::clicked, this, &QDialog::reject);
+
+  connect(co_compl, QOverload<const QString&>::of(&QCompleter::activated),
+          [=](const QString& text) {
+            combo_skylist->setCurrentText(text);
+            combo_skylist->setCurrentIndex(combo_skylist->findText(text));
+          });
+}
+
+QString CreateSkylanderDialog::GetFilePath() const
+{
+  return m_file_path;
+}
+
+void SkylanderPortalWindow::EmulatePortal(bool emulate)
+{
+  Config::SetBaseOrCurrent(Config::MAIN_EMULATE_SKYLANDER_PORTAL, emulate);
+  m_group_skylanders->setVisible(emulate);
+}
+
+void SkylanderPortalWindow::CreateSkylander(u8 slot)
+{
+  CreateSkylanderDialog create_dlg(this);
+  if (create_dlg.exec() == CreateSkylanderDialog::Accepted)
+  {
+    LoadSkylanderPath(slot, create_dlg.GetFilePath());
+  }
+}
+
+void SkylanderPortalWindow::LoadSkylander(u8 slot)
+{
+  const QString file_path = DolphinFileDialog::getOpenFileName(
+      this, tr("Select Skylander File"), m_last_skylander_path, tr("Skylander (*.sky);;"));
+  if (file_path.isEmpty())
+  {
+    return;
+  }
+  m_last_skylander_path = QFileInfo(file_path).absolutePath() + QString::fromStdString("/");
+
+  LoadSkylanderPath(slot, file_path);
+}
+
+void SkylanderPortalWindow::LoadSkylanderPath(u8 slot, const QString& path)
+{
+  File::IOFile sky_file(path.toStdString(), "r+b");
+  if (!sky_file)
+  {
+    QMessageBox::warning(
+        this, tr("Failed to open the skylander file!"),
+        tr("Failed to open the skylander file(%1)!\nFile may already be in use on the portal.")
+            .arg(path),
+        QMessageBox::Ok);
+    return;
+  }
+  std::array<u8, 0x40 * 0x10> file_data;
+  if (!sky_file.ReadBytes(file_data.data(), file_data.size()))
+  {
+    QMessageBox::warning(
+        this, tr("Failed to read the skylander file!"),
+        tr("Failed to read the skylander file(%1)!\nFile was too small.").arg(path),
+        QMessageBox::Ok);
+    return;
+  }
+
+  ClearSkylander(slot);
+
+  u16 sky_id = file_data[0x11];
+  u16 sky_var = file_data[0x1D];
+  sky_id <<= 8;
+  sky_var <<= 8;
+  sky_id |= file_data[0x10];
+  sky_var |= file_data[0x1C];
+
+  DEBUG_LOG_FMT(IOS_USB, "Sky Id: {}, 0x10: {} 0x11: {}", sky_id, file_data[0x10], file_data[0x11]);
+  DEBUG_LOG_FMT(IOS_USB, "Sky Var: {}, 0x1D: {} 0x1C: {}", sky_var, file_data[0x1D],
+                file_data[0x1C]);
+
+  u8 portal_slot = IOS::HLE::USB::g_skyportal.LoadSkylander(file_data.data(), std::move(sky_file));
+  sky_slots.at(slot) = {portal_slot, sky_id, sky_var};
+  UpdateEdits();
+}
+
+void SkylanderPortalWindow::ClearSkylander(u8 slot)
+{
+  if (auto slot_infos = sky_slots.at(slot))
+  {
+    auto [cur_slot, id, var] = slot_infos.value();
```

Comment:
Now that slot_infos is a proper struct and not just a tuple, you might as well remove this and just do `slot_info.portal_slot` on the line below.